### PR TITLE
ci: Use Robolectric instead of emulator

### LINKED ISSUE #645: Reduce Travis matrix
**Is your feature request related to a problem? Please describe.**
Android builds are taking up quite a bit of compute time and resources on Travis which slows down jobs for other repos in our org. 

 **Describe the solution you'd like**
We should come up with another approach to testing across device configurations that doesn't block Travis. Perhaps try out https://firebase.google.com/docs/test-lab

 **Describe alternatives you've considered**
N/A

 **Additional context**
N/A

cc @jpoehnelt

### PR DESCRIPTION
**Pull request by barbeau**
~~**Please don't merge - work-in-progress**~~

This is an effort to get Robolectric up and running.

Here's the current configuration:
* Move all tests from `androidTest` folder to `test` - it seems this should work, as Robolectric config docs say to use `testImplementation` (and not `androidTestImplementation`)
* Use `@RunWith(RobolectricTestRunner.class)` on the tests formerly in `androidTest` folder
* Use Robolectric `ShadowInstrumentation` instead of Android `InstrumentationRegistry`
* ~~Multidex is enabled (for now, to get around method limit - not sure if this is needed yet)~~ ( I've removed it - seems it was needed before the config to exclude all the maven dependencies)

I'm currently hitting a build failure on the test resources in the `raw` directory (`R.raw`), which is now under the `test` folder (see screenshot below).

The Robolectric docs seem to indicate that including `includeAndroidResources = true` in `build.gradle` should resolve this, but it's not :slightly_frowning_face:. I've also tried manually configuring `sourceSets` in `build.gradle` to include the `test/res` directory (currently commented out, as it shouldn't be needed), but that doesn't change anything either.

@arriolac Do you see anything obvious I'm missing for generally setting this up normally on Android?

I'll keep plugging away...

TODO:
- [x] Fix build errors
- [x] Get Roboelectric tests to run
- [x] Update Travis script - remove execution of instrumented tests

image

Fixes #645

### CODE REVIEW

**Comment by barbeau** [comment]
Some progress (I think). I've managed to get the tests to compile with the following changes:
* Switch to 4.3.1 and configure running on JDK 8 using properties file (apparently this is possible - see https://stackoverflow.com/a/59649406/937715)
* Move resource to normal test resources folder (not Android resources)
* Try to load resources using normal FileInputStream instead of Android-specific ShadowInstrumentation

However, if I try to run the `KmlContainerParserTest` from Android Studio, I see this in the log and then execution just hangs indefinitely:

~~~
[Robolectric] com.google.maps.android.data.kml.KmlContainerParserTest.testCreateContainerPlacemark: sdk=28; resources=BINARY
Called loadFromPath(/system/framework/framework-res.apk, true); mode=binary sdk=28
~~~

If I run `gradlew test` from the command line I see this:

~~~
E:\Android_Studio\android-maps-utils>gradlew test
Starting a Gradle Daemon, 1 incompatible Daemon could not be reused, use --status for details
Configuration on demand is an incubating feature.

> Task :library:compileDebugJavaWithJavac
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.

> Task :demo:compileDebugJavaWithJavac
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: E:\Android_Studio\android-maps-utils\demo\src\main\java\com\google\maps\android\utils\demo\PolySimplifyDemoActivity.java uses unchecked or unsaf
e operations.
Note: Recompile with -Xlint:unchecked for details.

> Task :library:javaPreCompileDebugUnitTest
Annotation processors must be explicitly declared now.  The following dependencies on the compile classpath are found to contain annotation processor.
  Please add them to the testAnnotationProcessor configuration.
  - auto-service-1.0-rc4.jar (com.google.auto.service:auto-service:1.0-rc4)
Alternatively, set android.defaultConfig.javaCompileOptions.annotationProcessorOptions.includeCompileClasspath = true to continue with previous behavi
or.  Note that this option is deprecated and will be removed in the future.
See https://developer.android.com/r/tools/annotation-processor-error-message.html for more details.

> Task :library:compileDebugUnitTestJavaWithJavac
Note: E:\Android_Studio\android-maps-utils\library\src\test\java\com\google\maps\android\data\kml\KmlFeatureParserTest.java uses unchecked or unsafe o
perations.
Note: Recompile with -Xlint:unchecked for details.

> Task :library:testDebugUnitTest
java.lang.NoClassDefFoundError: jdk/internal/reflect/GeneratedSerializationConstructorAccessor1
        at jdk.internal.reflect.GeneratedSerializationConstructorAccessor1.newInstance(Unknown Source)
        at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
        at java.base/java.io.ObjectStreamClass.newInstance(ObjectStreamClass.java:1081)
        at java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2071)
        at java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1594)
        at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:430)
        at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:121)
        at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:69)
        at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:68)
        at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:73)
Caused by: java.lang.ClassNotFoundException: jdk.internal.reflect.GeneratedSerializationConstructorAccessor1
        at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:471)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:588)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
        ... 10 more
java.lang.NoClassDefFoundError: jdk/internal/reflect/GeneratedSerializationConstructorAccessor1
        at jdk.internal.reflect.GeneratedSerializationConstructorAccessor1.newInstance(Unknown Source)
        at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
        at java.base/java.io.ObjectStreamClass.newInstance(ObjectStreamClass.java:1081)
        at java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2071)
        at java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1594)
        at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:430)
        at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:121)
        at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:69)
        at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:68)
        at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:73)
Caused by: java.lang.ClassNotFoundException: jdk.internal.reflect.GeneratedSerializationConstructorAccessor1
        at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:471)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:588)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
        ... 10 more
Process 'Gradle Test Executor 2' finished with non-zero exit value 1
org.gradle.process.internal.ExecException: Process 'Gradle Test Executor 2' finished with non-zero exit value 1
        at org.gradle.process.internal.DefaultExecHandle$ExecResultImpl.assertNormalExitValue(DefaultExecHandle.java:409)
        at org.gradle.process.internal.worker.DefaultWorkerProcess.onProcessStop(DefaultWorkerProcess.java:141)
        at org.gradle.process.internal.worker.DefaultWorkerProcess.access$000(DefaultWorkerProcess.java:42)
        at org.gradle.process.internal.worker.DefaultWorkerProcess$1.executionFinished(DefaultWorkerProcess.java:94)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
        at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:42)
        at org.gradle.internal.event.BroadcastDispatch$SingletonDispatch.dispatch(BroadcastDispatch.java:231)
        at org.gradle.internal.event.BroadcastDispatch$SingletonDispatch.dispatch(BroadcastDispatch.java:150)
        at org.gradle.internal.event.ListenerBroadcast.dispatch(ListenerBroadcast.java:141)
        at org.gradle.internal.event.ListenerBroadcast.dispatch(ListenerBroadcast.java:37)
        at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
        at com.sun.proxy.$Proxy101.executionFinished(Unknown Source)
        at org.gradle.process.internal.DefaultExecHandle.setEndStateInfo(DefaultExecHandle.java:217)
        at org.gradle.process.internal.DefaultExecHandle.finished(DefaultExecHandle.java:349)
        at org.gradle.process.internal.ExecHandleRunner.completed(ExecHandleRunner.java:110)
        at org.gradle.process.internal.ExecHandleRunner.run(ExecHandleRunner.java:84)
        at org.gradle.internal.operations.CurrentBuildOperationPreservingRunnable.run(CurrentBuildOperationPreservingRunnable.java:42)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
        at java.base/java.lang.Thread.run(Thread.java:834)

> Task :demo:compileReleaseJavaWithJavac
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: E:\Android_Studio\android-maps-utils\demo\src\main\java\com\google\maps\android\utils\demo\PolySimplifyDemoActivity.java uses unchecked or unsaf
e operations.
Note: Recompile with -Xlint:unchecked for details.
...
~~~

...and a bunch more `java.lang.NoClassDefFoundError: jdk/internal/reflect/GeneratedSerializationConstructorAccessor1`.

**Comment by barbeau** [comment]
Looks like this above crash was an issue with Jacoco config. See:
* https://github.com/gradle/gradle/issues/5184#issuecomment-457865951

After changing that config, now I'm seeing failures for the Robolectric tests:

~~~
> Task :library:testDebugUnitTest
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.robolectric.util.ReflectionHelpers (file:/C:/Users/barbeau/.gradle/caches/modules-2/files-2.1/org.robolectri
c/shadowapi/4.3.1/a93e55b36c418aa3afc17c26eac82590a3bfb63e/shadowapi-4.3.1.jar) to field java.lang.reflect.Field.modifiers
WARNING: Please consider reporting this to the maintainers of org.robolectric.util.ReflectionHelpers
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release

com.google.maps.android.data.geojson.GeoJsonFeatureTest > testGeometry FAILED
    org.apache.tools.ant.BuildException
        Caused by: org.apache.maven.artifact.resolver.MultipleArtifactsNotFoundException
~~~

**Comment by barbeau** [comment]
Ok, with the last commit (https://github.com/googlemaps/android-maps-utils/pull/649/commits/19cb9b0e8793399e8de0fe7b0074b54d43128e62) it looks like Roboelectric runs for most of the tests that were previously instrumented - here were the changes:
* Return default values for tests
* Remove test runner annotation from utility classes
* Ignore and annotate tests that currently don't work

Here are the rough stats for before/after Robolectric for running a build on Travis:

  | Emulator | Robolectric
-- | -- | --
~Each API level (minutes) | 9 | 3
~Total for 12 API levels (minutes) | 26 | 10.5

Obviously some of the API level builds run in parallel. So total execution time is roughly a third of what it was previously 🎉.

@arriolac This PR should be review for review now - let me know if you want to trim any API levels from the build to further cut down time.

File: library/src/test/resources/robolectric.properties [hunk 1]
```diff
@@ -0,0 +1 @@
+sdk=28
```
Comment (arriolac) [inline library/src/test/resources/robolectric.properties]: Should this be 29? `compileSdkVersion` and `targetSdkVersion` are both set to that.

Comment (barbeau) [reply library/src/test/resources/robolectric.properties]: This has to be `28` to run on Java 8. If it's `29`, it forces use of Java 9, which currently requires manual configuration in Android Studio to run - see https://stackoverflow.com/questions/56821193/does-robolectric-require-java-9/59649406#59649406.

**Pull request merged**